src/devices/Mcp3008/Mcp3008.cs

```diff
@@ -56,88 +60,102 @@ public void Dispose()
             }
         }
 
-        public int Read(int adc_channel)
+        public int Read(int channel, InputConfiguration inputConfiguration = InputConfiguration.SingleEnded)
         {
-            if (adc_channel < 0 || adc_channel > 7)
+            if (channel < 0 || channel > 7)
             {
                 throw new ArgumentException("ADC channel must be within 0-7 range.");
             }
 
             if (_protocol == CommunicationProtocol.Spi)
             {
-                return ReadSpi(adc_channel);
+                return ReadSpi(channel, inputConfiguration);
             }
             else
             {
-                return ReadGpio(adc_channel);
+                return ReadGpio(channel, inputConfiguration);
             }
         }
 
+        private static byte GetConfigurationBits(int channel,InputConfiguration inputConfiguration = InputConfiguration.SingleEnded)
+        {
+            int configurationBits;
+
+            if (inputConfiguration == InputConfiguration.SingleEnded)
+            {
+                configurationBits = 0b00011000 | channel;
+            }
+            else
+            {
+                configurationBits = 0b00010000 | channel;
+            }
+
+            configurationBits = configurationBits << 3;
+            return (byte)configurationBits;
+        }
+
         // port of https://gist.github.com/ladyada/3151375
-        private int ReadGpio(int adc_channel)
+        private int ReadGpio(int channel, InputConfiguration inputConfiguration)
         {
             while (true)
             {
-                var commandout = 0;
-                var trim_pot = 0;
-
-                _controller.Write(_CS, PinValue.High);
-                _controller.Write(_CLK, PinValue.Low);
-                _controller.Write(_CS, PinValue.Low);
+                var result = 0;
+                byte command = GetConfigurationBits(channel, inputConfiguration);
```

Comment:
Correct me if I’m wrong but this one won’t be the same as 0x18 << 3 right? Since your method would do (0x18 | channel) << 3

Reply:
@joperezr You are correct, but this is the bug fix when reading a channel using GPIO.  Notice in original code the `adc_channel` parameter is not used and command is basically hard-coded to only read channel 0.  This helper method also allows for single-ended or differential mode, as well.